pkg/controller/fileobserver/observer_polling.go

```diff
@@ -0,0 +1,123 @@
+package fileobserver
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type pollingObserver struct {
+	interval time.Duration
+	reactors map[string][]reactorFn
+	files    map[string]string
+
+	reactorsMutex sync.RWMutex
+}
+
+// AddReactor will add new reactor to this observer.
+func (o *pollingObserver) AddReactor(reaction reactorFn, files ...string) Observer {
+	o.reactorsMutex.Lock()
+	defer o.reactorsMutex.Unlock()
+	for _, f := range files {
+		// Do not rehash existing files
+		if _, exists := o.files[f]; exists {
+			continue
+		}
+		var err error
+		glog.V(3).Infof("Adding reactor for file %q", f)
+		o.files[f], err = calculateFileHash(f)
+		if err != nil {
+			panic(fmt.Sprintf("unexpected error while adding reactor for %#v: %v", files, err))
+		}
+		o.reactors[f] = append(o.reactors[f], reaction)
+	}
+	return o
+}
+
+func (o *pollingObserver) processReactors(stopCh <-chan struct{}) {
+	err := wait.PollImmediateInfinite(o.interval, func() (bool, error) {
+		select {
+		case <-stopCh:
+			return true, nil
+		default:
+		}
+		for filename, reactors := range o.reactors {
```

Comment:
you need a lock here

Reply:
that will mean you won't be able to add new reactors after you run, right?

Reply:
iow. it will block until all reactors are processed.